examples/dfem/minimal_surface.cpp

```diff
@@ -0,0 +1,518 @@
+//               in one of three ways (controlled by -der command line parameter):
+//               0) Automatic differentiation using Enzyme or dual type (default)
+//               1) Hand-coded derivatives
+//               2) Finite differences
+//
+//               The example demonstrates the use of MFEM's nonlinear solvers,
+//               automatic differentiation capabilities, and GLVis/ParaView visualization.
+
+#include "mfem.hpp"
+
+using namespace mfem;
+
+// This example code demonstrates the use of new features in MFEM that are in
+// development but exposed through the mfem::future namespace. All features
+// under this namespace might change their interface or behavior in upcoming
+// releases until they have stabilized.
+using namespace mfem::future;
+using mfem::future::tensor;
+
+// Derivative type enum
+// This enum is used to specify the type of derivative computation. Possibilities are
+// AUTODIFF, which uses automatic differentiation (Enzyme or dual type)
+// HANDCODED, which uses a manually implemented derivative and
+// FD (finite difference).
+enum DerivativeType
+{
+   AUTODIFF,
+   HANDCODED,
+   FD
+};
+
+// Minimal surface operator.
+//
+// This class implements the minimal surface equation, which is a nonlinear
+// operator that provides the residual.
+template <typename dscalar_t, int dim = 2>
+class MinimalSurface : public Operator
+{
+private:
+   static constexpr int SOLUTION_U = 1;
+   static constexpr int MESH_NODES = 2;
+
+   template <typename T>
+   MFEM_HOST_DEVICE inline
+   static auto coeff(const tensor<T, dim> &a)
+   {
+      return 1_r / sqrt(1_r + sqnorm(a));
+   }
+
+   // Matrix-Free version of the pointwise residual form for the minimal
+   // surface equation.
+   struct MFApply
+   {
+      // Using DifferentiableOperator, we can define the residual form as a
+      // matrix-free operation. This allows us to compute the residual without
+      // explicitly forming any matrices or other large, temporary data
+      // structures.
+      //
+      // The inputs are the gradient of the solution in *reference coordinates*,
+      // the Jacobian of the coordinates, and the integration rule weights.
+      //
+      // The output is the residual in *physical coordinates* which also
+      // includes the necessary transformation from reference to physical
+      // coordinates for the gradient of the test function.
+      //
+      // Due to the description of how this pointwise operation is used in
+      // DifferentiableOperator, we know it is applied to the gradient of the
+      // test function in reference coordinates e.g.
+      // $ \int coeff(\nabla_x u) (\nabla_x u) J^{-T} \det(J) w \nabla_{\xi} v d\xi$
+      MFEM_HOST_DEVICE inline
+      auto operator()(
+         const tensor<dscalar_t, dim> &dudxi,
+         const tensor<real_t, dim, dim> &J,
+         const real_t &w) const
+      {
+         const auto invJ = inv(J);
```

Comment:
auto is hiding the outputs - for new users it will be better to see the actual types